log/json_logger.go

```diff
@@ -1,91 +1,15 @@
 package log
 
 import (
-	"encoding"
-	"encoding/json"
-	"fmt"
 	"io"
-	"reflect"
-)
 
-type jsonLogger struct {
-	io.Writer
-}
+	"github.com/go-kit/log"
+)
 
 // NewJSONLogger returns a Logger that encodes keyvals to the Writer as a
 // single JSON object. Each log event produces no more than one call to
 // w.Write. The passed Writer must be safe for concurrent use by multiple
 // goroutines if the returned Logger will be used concurrently.
 func NewJSONLogger(w io.Writer) Logger {
-	return &jsonLogger{w}
-}
-
-func (l *jsonLogger) Log(keyvals ...interface{}) error {
-	n := (len(keyvals) + 1) / 2 // +1 to handle case when len is odd
-	m := make(map[string]interface{}, n)
-	for i := 0; i < len(keyvals); i += 2 {
-		k := keyvals[i]
-		var v interface{} = ErrMissingValue
-		if i+1 < len(keyvals) {
-			v = keyvals[i+1]
-		}
-		merge(m, k, v)
-	}
-	enc := json.NewEncoder(l.Writer)
-	enc.SetEscapeHTML(false)
-	return enc.Encode(m)
-}
-
-func merge(dst map[string]interface{}, k, v interface{}) {
-	var key string
-	switch x := k.(type) {
-	case string:
-		key = x
-	case fmt.Stringer:
-		key = safeString(x)
-	default:
-		key = fmt.Sprint(x)
-	}
-
-	// We want json.Marshaler and encoding.TextMarshaller to take priority over
-	// err.Error() and v.String(). But json.Marshall (called later) does that by
-	// default so we force a no-op if it's one of those 2 case.
-	switch x := v.(type) {
-	case json.Marshaler:
-	case encoding.TextMarshaler:
-	case error:
-		v = safeError(x)
-	case fmt.Stringer:
-		v = safeString(x)
-	}
-
-	dst[key] = v
-}
-
-func safeString(str fmt.Stringer) (s string) {
-	defer func() {
-		if panicVal := recover(); panicVal != nil {
-			if v := reflect.ValueOf(str); v.Kind() == reflect.Ptr && v.IsNil() {
-				s = "NULL"
-			} else {
-				panic(panicVal)
-			}
-		}
-	}()
-	s = str.String()
-	return
-}
-
-func safeError(err error) (s interface{}) {
-	defer func() {
-		if panicVal := recover(); panicVal != nil {
-			if v := reflect.ValueOf(err); v.Kind() == reflect.Ptr && v.IsNil() {
-				s = nil
-			} else {
-				panic(panicVal)
-			}
-		}
-	}()
-	s = err.Error()
-	return
+	return log.NewJSONLogger(w)
```

Comment:
Is this the preferred way to point to the new module? Versus e.g.

```go
var NewJSONLogger = log.NewJSONLogger
```

Reply:
Good question. I had to think about that a bit before responding.

I dug back through the alias proposal to refresh my memory on the reason why it only needed to worry about types and not other kinds of declarations. A forwarding function is the right approach according to the [motivating article](https://talks.golang.org/2016/refactor.article#TOC_5).

I also thought this through from first principles a bit and I have two arguments for using a forwarding function.

1. The packages docs for the legacy package functions will continue to render with full function signatures. That seems better than rendering as variables.
2. As of [Go 1.12, functions that only call another function are inlined](https://golang.org/doc/go1.12#compiler). That should mitigate any concerns about call overhead. Furthermore, using a variable instead defeats compiler inlining in `kit/log` because the compiler cannot be certain about what function will be stored in that variable from call to call. Uses of the variable are indirect function calls and cannot be inlined.

I have verified that using a forwarding function gets inlined as we would hope. Building `examples/stringsvc2` that imports  `kit/log` with the code in this PR I see:

```
.\main.go:16:31: inlining call to "github.com/go-kit/kit/log".NewLogfmtLogger      
.\main.go:16:31: inlining call to "github.com/go-kit/log".NewLogfmtLogger
```

If I switch to a function variable for `kit/log.NewLogfmtLogger` it does not get inlined.

Reply:
I'm sold.